### ISSUE TITLE
[Security] backdrop_http_request() fails to strip Cookie or Authorization headers on HTTP downgrade

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#5963]
This is cross-port of the D7 fix [e15ccfbf6a](https://git.drupalcode.org/project/drupal/-/commit/e15ccfbf6a56f071f66eff9183a5a9f9dcf666dd) | [Issue #3293649 by mcdruid: drupal_http_request() fails to strip Cookie or Authorization headers on HTTP downgrade](https://drupal.org/node/3293649). It's tracked by the task list #5770.

Adapting from the description from D7:

Backdrop 1.x has a vulnerability similar to the two recently fixed in Guzzle:

- https://github.com/guzzle/guzzle/security/advisories/GHSA-w248-ffj2-4v5q
- https://github.com/guzzle/guzzle/security/advisories/GHSA-f2wf-25xc-69c9

backdrop_http_request() automatically follows 3xx redirects. RFC 9110 states:

> When automatically following a redirected request, the user agent SHOULD resend the original request message with the following modifications:
> 
> 3. Consider removing header fields that were not automatically generated by the implementation (i.e., those present in the request because they were added by the calling context) where there are security implications; this includes but is not limited to Authorization and Cookie.

See https://www.rfc-editor.org/rfc/rfc9110.html#name-redirection-3xx

This issue was discussed by the D7 security team, and it was decided that this can be handled as a security hardening issue in the public issue queue.

**Comment by argiepiano** [comment]
PR [4330](https://github.com/backdrop/backdrop/pull/4330)

The PR does the following:
- It adds function `_backdrop_should_strip_sensitive_headers_on_http_redirect()` that checks whether the redirect is a downgrade. If so, it returns true
- It strips the Headers `Cookie` and `Authorization` if needed (i.e. when the redirect has an HTTP downgrade)
- It adds a test that checks that this new function works correctly
- It adds a two configurations to `settings.php` that allows site admins to turn off the automatic stripping

**Comment by argiepiano** [comment]
Suggested change record:

By default `backdrop_http_request()` will strip sensitive request headers (Cookie, Authorization) when following a redirect if the redirect location has a different http host to the original request, or if the scheme downgrades from https to http.

`drupal_http_request()` automatically follows 3xx redirects. RFC 9110 states:

>When automatically following a redirected request, the user agent SHOULD resend the original request message with the following modifications:

>3. Consider removing header fields that were not automatically generated by the implementation (i.e., those present in the request because they were added by the calling context) where there are security implications; this includes but is not limited to Authorization and Cookie.

See https://www.rfc-editor.org/rfc/rfc9110.html#name-redirection-3xx

Sites can opt-out from this behavior by setting the drupal_http_request_strip_sensitive_headers_on_host_change or drupal_http_request_strip_sensitive_headers_on_https_downgrade variable in settings.php:

```php
/**
 * Sensitive request headers in backdrop_http_request() when following a
 * redirect.
 *
 * By default backdrop_http_request() will strip sensitive request headers when
 * following a redirect if the redirect location has a different http host to
 * the original request, or if the scheme downgrades from https to http.
 *
 * These variables allow opting out of this behaviour. Careful consideration of
 * the security implications of opting out is recommended. To opt out, set to
 * FALSE.
 *
 * @see _backdrop_should_strip_sensitive_headers_on_http_redirect()
 * @see backdrop_http_request()
 */
// $config['backdrop_http_request_strip_sensitive_headers_on_host_change'] = TRUE;
// $config['backdrop_http_request_strip_sensitive_headers_on_https_downgrade'] = TRUE;
```

**Comment by klonos** [comment]
Thanks @argiepiano ...I've left some comments/questions and code suggestions in the PR.

**Comment by argiepiano** [comment]
Thanks, @klonos. Good catches! I've pushed those changes. I'm getting a strange coding standard error - `/bin/sh: 1: phpcs: not found`. Any clue what's going on?

**Comment by klonos** [comment]
Code looks good @argiepiano 👍🏼 ...that error is random. It usually goes away when I rebase my PRs or if I close/wait/reopen.

**Comment by BWPanda** [comment]
I'm wondering about the need for such verbose function/settings names... Could we try to shorten them somewhat?

```
function _check_strip_sensitive_headers($url, $location) {
}

$settings['strip_sensitive_headers_on_host_change'] = TRUE;
$settings['strip_sensitive_headers_on_https_downgrade'] = TRUE;
```

**Comment by klonos** [comment]
@BWPanda I had the same thought, however this is what these things have been named in D7. So if we rename them in Backdrop (beyond the `drupal_` prefix that is), we may make it slightly more complicated when porting things from D7 (for instance, I imagine that modules like coder_upgrade would then need to keep track of these via a mapping array or something, instead of straight conversion from `$conf['drupal_xyz']` to `$settings['backdrop_xyz']`). So 🤔

**Comment by quicksketch** [comment]
A few small things in the PR:

* There's a `drupal_` namespace that slipped in: https://github.com/backdrop/backdrop/pull/4330/files#r1136511795

**Comment by BWPanda** [comment]
> Or we group them in the event we have other config settings for backdrop_http_request in the future:
>
> ```
> $config['system.core']['backdrop_http_request']['strip_sensitive_headers_on_host_change']
> $config['system.core']['backdrop_http_request']['strip_sensitive_headers_on_https_downgrade']
> ```

Yeah, I prefer this.

**Comment by argiepiano** [comment]
@quicksketch apologies for the long time making the requested changes. This fell through the cracks.

I've made the changes you requested - namely using config instead of settings for the new flags. The coding standard failures are mostly things that this PR did not touch. One of those things that this PR DID add is a commented line in settings.php that is too long - but that comment is an example of what the $config would look like if overridden, so it doesn't make sense to split it (?).

Ready for code review again.

**Comment by argiepiano** [comment]
Well it seems the new automated test is failing (but it is passing in my local, so I can't explain what's going on). I'll have to look at this later.

**Comment by argiepiano** [comment]
OK, tests are passing now (except for the coding standard things I mentioned above)

**Comment by herbdool** [comment]
It's been rebased. The code matches the D7 commit, which also included full testing coverage. This would be hard to manually test, but since it's already in D7 and because it's a hardening, I think we should merge.

**Comment by laryn** [comment]
Thanks for cross-porting this @argiepiano, and for the suggestion to use config overrides instead of settings @quicksketch. Also thanks for reviewing and feedback @herbdool, @klonos and Peter Anderson.
With thanks also to mcdruid for the work in Drupal 7.
Merged into `1.x` and `1.27.x`.

**Comment by klonos** [comment]
Keeping open for the change record to be created.

**Comment by klonos** [comment]
Follow-up PR: https://github.com/backdrop/backdrop/pull/4753

I'll also try to draft a change record ...unless someone else beats me to it if I get distracted by other things.

**Comment by argiepiano** [comment]
LGTM. 

Change record: weird situation. I don't have login credentials for `https://docs.backdropcms.org/`. When I try to create an account I get to reset password, I get `The name argiepiano is already taken.
The email address argiepiano@yahoo.com is already registered`, and when I try to reset password I get: `Sorry, argiepiano@yahoo.com is not recognized as a user name or an email address.`  👎🏽 

@klonos, any change you can check on that?

**Comment by bugfolder** [comment]
User account @argiepiano was blocked. (It was a year old, never accessed.) I've unblocked it, so password recovery should work now.

(Perhaps the opacity of the error message might be for reasons of security, but it seems like if an account is blocked, we ought to say so.)

**Comment by argiepiano** [comment]
Here is a draft. I selected markdown for the editor, but for some reason it's adding links to all my formatted text. Don't know how to fix that.

https://docs.backdropcms.org/change-records/backdrop_http_request-now-strips-cookie-and-authorization-headers-on-http-downgrade-or-http-host-change

**Comment by bugfolder** [comment]
>  for some reason it's adding links to all my formatted text. Don't know how to fix that.

On the docs site, all of the text formats, including Markdown, include the **API filter**, in which "Function names (and classes, constants, etc.) turn into links."

**Comment by laryn** [comment]
Note: removing the RTBC because this has already been committed, just needs the change record to be approved.
